src/lazy/expanded/macro_table.rs

```diff
@@ -171,23 +174,17 @@ impl<'top> MacroRef<'top> {
 #[derive(Debug, Clone)]
 pub struct MacroTable {
     // Stores `Rc` references to the macro definitions to make cloning the table's contents cheaper.
-    macros_by_address: Vec<Rc<Macro>>,
+    macros_by_address: Vec<Arc<Macro>>,
     // Maps names to an address that can be used to query the Vec above.
-    macros_by_name: FxHashMap<Rc<str>, usize>,
+    macros_by_name: FxHashMap<CompactString, usize>,
 }
 
-thread_local! {
-    /// An instance of the Ion 1.1 system macro table is lazily instantiated once per thread,
-    /// minimizing the number of times macro compilation occurs.
-    ///
-    /// The thread-local instance holds `Rc` references to its macro names and macro definitions,
-    /// making its contents inexpensive to `clone()` and reducing the number of duplicate `String`s
-    /// being allocated over time.
-    ///
-    /// Each time the user constructs a reader, it clones the thread-local copy of the system macro
-    /// table.
-    pub static ION_1_1_SYSTEM_MACROS: MacroTable = MacroTable::construct_system_macro_table();
-}
+/// A lazily initialized singleton instance of the Ion 1.1 system macro table.
+///
+/// This instance is shared by all readers, minimizing the number of times that macro compilation
+/// and heap allocation occurs.
+pub static ION_1_1_SYSTEM_MACROS: LazyLock<MacroTable> =
+    LazyLock::new(MacroTable::construct_system_macro_table);
```

Comment:
🗺️ Originally, each reader or writer would instantiate its own copy of the system macro table. This involved several heap allocations and some compilation logic that could feel quite heavyweight for a short-lived stream.

To avoid paying this cost for every reader and writer, I made the system macro table a thread-local value. Only the first reader or writer on a thread would pay the initialization cost, and `Macro` and `Parameter` instances could continue using lightweight `Rc`s to share data within the thread.

However, this introduced a problem: unlike `static` values that live for the duration of the program, thread locals only live for the duration of their thread; they do not have a named lifetime.  You can only access a thread local value [via a closure](https://doc.rust-lang.org/std/thread/struct.LocalKey.html#method.with), and you cannot retain a reference to it. As a result, it wasn't possible to get a reference (`&`) to a system macro from the thread local system macro table—only an `Rc`.

As a compromise, I used a thread local to build the first system macro table—performing the necessary macro compilation steps—and then had subsequent readers and writers `clone()` that instance of the table. This allowed most readers and writers to avoid the cost of compiling system macros, but they still had to pay for the heap allocations. Not terribly satisfying.

Recently, in Rust v1.80, [the `LazyLock` type](https://doc.rust-lang.org/beta/std/sync/struct.LazyLock.html) was stabilized. It uses synchronization primitives to manage initializing the value, but all accesses thereafter can get a `&` reference for 'free'.

By switching the internals of the `Macro` type from `Rc<str>` to `CompactString` and `Rc<Macro>` to `Arc<Macro>`, I was able to create an immutable, lazily-initialized global singleton system macro table. References to the system macros are now trivial to get and can be reused across all readers and writers.